### ISSUE TITLE
Segmentation Fault running script (recursive function) - Linux x64

### ISSUE THREAD

**Issue by magonzalez192** [sol-lang/sol#5]
Hi,

I'm getting `segmentation fault` error when execute this recursive function:

```
func loop(i) 
    a = i
    while a > 0 do
        a = a - 1
        loop(a - 1)
    end
    return a
end

print("loop value: ", loop(100000))
```

I'll check valgrind for look what happend, but it's also happend with simple recursive function.

**Comment by Grissess** [comment]
Confirmed on my laptop (same arch); the given program crashes due to stack overflow (53906 C stack frames).

The underlying cause is twofold: Sol executes and evaluates programs by directly traversing the parse tree recursively (causing already undue C stack usage), and it is at least hard to implement tail recursion due to dynamic scoping (it's impossible to tell ahead of time what a name will resolve to in a given scope--it's even possible to hijack the process from within user code).

The "easy" resolution (though perhaps cheating) would be to grow the C stack; the more involved solution would be to properly linearize code so that recursion depends on a VM's stack instead of the C run-time stack. Both of these are somewhat involved, and the former is not particularly portable nor reliable. The latter risks making breaking changes to the current semantics.

For now, I would recommend iterative algorithms over recursive algorithms, if at all possible. If not, try increasing (native) stack space of the running application (e.g., `ulimit -s`).

**Comment by magonzalez192** [comment]
So a way for fix it would be implement tail recursion? (some in mind...)

**Comment by Grissess** [comment]
After some serious thought, I've come to the conclusion that tail recursion may be possible in the following circumstances:
- The return expression in an `ST_RET` is exactly an `EX_CALL`.
- When evaluated, the  `expr` of the `EX_CALL` is a function object AND the AST node of that function (`func`) is exactly (pointer-) equivalent. This is necessary because [function ASTs can be changed at runtime](https://github.com/Grissess/sol/blob/0cea7dff12ad015f0d376c80a19e67eec532a69c/programs/test.sol#L217-L248).

Tail recursion can only approach correctness if these conditions are met, and--even then--information is lost in the traceback, because it will not be obvious from the stack of `sol_exec` calls that the function was evaluated more than once. Python's creator had [also](http://neopythonic.blogspot.com.au/2009/04/tail-recursion-elimination.html) [considered](http://neopythonic.blogspot.com.au/2009/04/final-words-on-tail-calls.html) implementing this, but chose not to for the latter reason. Nonetheless, modifying the calling convention(s) of Sol is generally as simple as changing [`sol_f_func_call`](https://github.com/Grissess/sol/blob/0cea7dff12ad015f0d376c80a19e67eec532a69c/runtime.c#L908-L958), for which a simple `do {...} while` might suffice in implementing that feature.

**Comment by Grissess** [comment]
As a workaround fix, TCO should happen starting in 84aed45 . Sol is *very* dynamic, so the runtime is only smart enough to do this for **return statements whose only expression is a call to the running function**. This _doesn't_ fix the case above, nor will it work for mutually recursive functions, but it seems to satisfy `tests/crasher_tco.sol` for now, after a minor rework from the original.

Getting the running function reliably can be done in a number of ways:
- If the function has a name (usually as given in the expression), this name is assigned the function value as the last operation on the scope before the function starts running (runtime.c:L1098 in my local copy). The name can be set by assigning to the `name` index of a function object (and read analogously);
- `debug.fnstack()` returns a reference to `state->fnstack`, which is a Sol list representing the current function call stack, latest first. (CFunction calls also exist in here, but the latest call *should* be a Sol function if Sol code inside a recursive function is running. The returned value is a mutable reference and this list is used for some runtime safety checks, so this method is a little dangerous.)

Ultimately, I'd be a bigger fan of a fix for #21, but this will hold over that particular crasher for now. I still recommend iterative solutions when feasible.
